source/extensions/filters/common/rbac/engine_impl.cc

```diff
@@ -26,10 +26,45 @@ RoleBasedAccessControlEngineImpl::RoleBasedAccessControlEngineImpl(
   }
 }
 
-bool RoleBasedAccessControlEngineImpl::allowed(const Network::Connection& connection,
-                                               const Envoy::Http::RequestHeaderMap& headers,
-                                               const StreamInfo::StreamInfo& info,
-                                               std::string* effective_policy_id) const {
+bool RoleBasedAccessControlEngineImpl::handleAction(const Network::Connection& connection,
+                                                    StreamInfo::StreamInfo& info,
+                                                    std::string* effective_policy_id) {
+  return handleAction(connection, *Http::StaticEmptyHeaders::get().request_headers, info,
+                      effective_policy_id);
+}
+
+bool RoleBasedAccessControlEngineImpl::handleAction(const Network::Connection& connection,
+                                                    const Envoy::Http::RequestHeaderMap& headers,
+                                                    StreamInfo::StreamInfo& info,
+                                                    std::string* effective_policy_id) {
+  bool matched = checkPolicyMatch(connection, info, headers, effective_policy_id);
+
+  switch (action_) {
+  case envoy::config::rbac::v3::RBAC::ALLOW:
+    return matched;
+  case envoy::config::rbac::v3::RBAC::DENY:
+    return !matched;
+  case envoy::config::rbac::v3::RBAC::LOG: {
+    // If not shadow enforcement, set shared log metadata
+    if (mode_ != EnforcementMode::Shadow) {
+      ProtobufWkt::Struct log_metadata;
+      auto& log_fields = *log_metadata.mutable_fields();
+      log_fields[DynamicMetadataKeysSingleton::get().AccessLogKey].set_bool_value(matched);
+      info.setDynamicMetadata(DynamicMetadataKeysSingleton::get().CommonNamespace, log_metadata);
+    }
+
+    return true;
+  }
+  default:
+    return true;
```

Comment:
This is checked by schema and should be NOT_REACHED

Reply:
case statements using proto_enums should have a default because there are other values added by the generator.